scripts/dev/tidy.php

```diff
@@ -0,0 +1,394 @@
+
+  DESCRIPTION:
+
+    Trims trailing whitespace (spaces and tabs) from each line and trims
+    multiple final new lines into a single new line. If file has different
+    line endings the prevailing EOL (end of line) character is used as a
+    final new line.
+
+    When used on the php-src root directory it ignores the predefined set
+    of binary files and certain test files.
+
+    By default, original files are backed up (file.ext~).
+
+  OPTIONS:
+
+    -h, --help           Display this help
+    -v, --version        Display version
+        --no-whitespace  Disable trimming trailing whitespace
+        --no-eol         Disable fixing final end of lines
+        --no-colors      Disable colors in the script output
+        --no-backup      Disable backed up files. Use with caution to not
+                         overwrite something by accident.
+    -f, --fix            Edit and save files in place
+    -q, --quiet          Do not output any message
+
+  EXAMPLES:
+
+    Display help: php tidy.php -h
+
+    Check php-src: scripts/dev/tidy.php .
+
+    Fix php-src: scripts/dev/tidy.php --fix --no-backup .
+
+    Fix custom path or file: ./tidy.php --fix input-path
+
+END;
+}
+
+/**
+ * Determine the script version. If script is located in the PHP repository the
+ * PHP version is returned, otherwise the script's file sha1 hash is returned.
+ */
+function version(): string
+{
+    $file = __DIR__.'/../../main/php_version.h';
+
+    if (file_exists($file)) {
+        $content = file_get_contents($file);
+        preg_match('/^\#define\ PHP\_VERSION\ \"([0-9\.\-a-z]+)\"$/m', $content, $version);
+
+        if (isset($version[1])) {
+            return $version[1];
+        }
+    }
+
+    return sha1_file(__FILE__);
+}
+
+/**
+ * Trim trailing spaces and tabs from each line.
+ */
+function trimTrailingWhitespace(string $content): string
+{
+    return preg_replace('/[\t ]+(\r?$)/m', '$1', $content);
+}
+
+/**
+ * Trim final new lines and append an EOL (end of line) based on the previous
+ * prevailing EOL type used LF (\n), CRLF (\r\n) or CR (\r).
+ */
+function trimFinalEol(string $content): string
+{
+    $trimmed = rtrim($content, " \t\r\n");
+    $trimmed .= getEol($trimmed);
+
+    return $trimmed;
+}
+
+/**
+ * Get end of line based on the prevailing LF, CRLF or CR new line characters.
+ */
+function getEol(string $content): ?string
+{
+    // Empty files don't need a new line attached
+    if (empty($content)) {
+        return null;
+    }
+
+    // Match all LF, CRLF and CR EOL characters
+    preg_match_all('/(*BSR_ANYCRLF)\R/', $content, $matches);
+
+    // For single line files LF is returned
+    if (empty($matches[0])) {
+        return "\n";
+    }
+
+    $counts = array_count_values($matches[0]);
+    arsort($counts);
+    $prevailingEol = key($counts);
+
+    return $prevailingEol;
+}
+
+$fix = false;
+$whitespace = true;
+$eol = true;
+$colors = true;
+$backup = true;
+$quiet = false;
+
+foreach ($argv as $argument) {
```

Comment:
Because of `register_argc_argv` ini setting, should you check availability of `$argv` before using it? Or maybe just `foreach ($argv ?? [] as $argument)`

Reply:
@webdevxp This script is meant to only ever be run as sapi/cli, which hardcodes register_argc_argv to always be true

Reply:
Thanks for the info... Yes, this script is actually meant to run as a CLI tool only.

Reply:
@KalleZ I've added this perversion to the deprecation list at https://wiki.php.net/rfc/deprecations_php_7_4#register_argc_argv_ini_directive. I'm wondering what the best approach for that would be, as the ini setting is unfortunately enabled by default on non-CLI SAPIs :/

Reply:
@KalleZ Sorry my bad. I misunderstood because I used to test with `php -dregister_argc_argv=0 -r 'var_dump($argv);'` resulted in `PHP Notice:  Undefined variable: argv in Command line code on line 1`. But I understand now it seems command line configs take precedence hardcoding.

Reply:
@nikic I think it might have been useful back in the old cgi-bin days (for whatever questionable reason) but it should really just be a CLI thing imo, seems like a nice catch for the list

Reply:
@webdevxp You can checkout the relevant code for CLI here: https://github.com/php/php-src/blob/master/sapi/cli/php_cli.c#L135